python/ray/experimental/channel/shared_memory_channel.py

```diff
@@ -480,6 +483,91 @@ def close(self) -> None:
         self._worker.core_worker.experimental_channel_set_error(self._reader_ref)
 
 
+@DeveloperAPI
+class BufferedSharedMemoryChannel(ChannelInterface):
+    """A Channel that allows to buffer inputs.
+
+    If the buffer is full for writes or empty for reads and the operation
+    blocks longer than configured timeouts, RayChannelTimeoutError will be
+    raised.
+
+    As long as the buffer is not full, read/write APIs are non-blocking.
```

Comment:
nit: let's be more accurate, for reads, it is non-blocking if the buffer is not empty.